manta-crypto/src/arkworks/glv/mod.rs

```diff
@@ -0,0 +1,341 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Arkworks Elliptic Curve Implementation
+
+use crate::arkworks::{
+    ec::{
+        models::{short_weierstrass_jacobian, SWModelParameters},
+        AffineCurve, ProjectiveCurve,
+    },
+    ff::PrimeField,
+};
+use alloc::{vec, vec::Vec};
+use ark_ff::{BigInteger, Field};
+use core::str::FromStr;
+use num_bigint::{BigInt, BigUint, Sign};
+
+#[cfg(feature = "ark-bls12-381")]
+use crate::arkworks::bls12_381;
+
+#[cfg(feature = "ark-bn254")]
+use crate::arkworks::bn254;
+
+/// Affine Curve Extension
+pub trait AffineCurveExt: AffineCurve {
+    /// Returns the `x` coordinate of `self`.
+    fn x(&self) -> &Self::BaseField;
+
+    /// Returns the `y` coordinate of `self`.
+    fn y(&self) -> &Self::BaseField;
+
+    /// Builds [`Self`] from `x` and `y`.
+    fn from_xy_unchecked(x: Self::BaseField, y: Self::BaseField) -> Self;
+
+    /// Applies the GLV endomorphism to `self`
+    #[inline]
+    fn glv_endomorphism(&self, beta: &Self::BaseField) -> Self {
+        Self::from_xy_unchecked(*self.x() * beta, *self.y())
+    }
+}
+
+impl<P> AffineCurveExt for short_weierstrass_jacobian::GroupAffine<P>
```

Comment:
I will because it can be useful to implement endomorphisms of Edwards curves in the future, but note that multiplying `x` by a cubic root of `1` is not a symmetry of an Edwards curve equation. It might be worth investigating the space of "efficient" endomorphisms on such curves to get something similar to GLV though.

Reply:
I see. In that case, I think we should move the `glv_endomrphism` function to the `HasGlv` trait instead of the `AffineCurveExt` trait.